src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -592,4 +594,31 @@ public void shutdown(int timeoutSeconds) {
     GreengrassService getMain() {
         return mainService;
     }
+
+    /**
+     * Check if all services has reached to terminal state: RUNNING, FINISHED or BROKEN.
+     *
+     */
+    public boolean allServicesInTerminalState() {
+        //1. Check all other component status
+        List<GreengrassService> servicesToTrack = kernel.orderedDependencies().stream()
+                .filter(GreengrassService::shouldAutoStart).collect(Collectors.toList());
+        boolean allReachedTerminalState = true;
+        for (GreengrassService service : servicesToTrack) {
+            State state = service.getState();
+            if (!service.reachedDesiredState()) {
+                allReachedTerminalState = false;
+                break;
+            }
+            if (State.RUNNING.equals(state) || State.FINISHED.equals(state) || !service.shouldAutoStart()
+                    && service.reachedDesiredState()) {
```

Comment:
`service.reachedDesiredState()`
i think your check on 609 covers this

Reply:
Updated